core/src/main/java/org/elasticsearch/transport/local/LocalTransport.java

```diff
@@ -288,6 +294,9 @@ private void handleRequest(StreamInput stream, long requestId, LocalTransport so
             if (reg == null) {
                 throw new ActionNotFoundTransportException("Action [" + action + "] not found");
             }
+            // reserve here in order be able to send a response in case we break the limit
+            inFlightRequestsBreaker.addEstimateBytesAndMaybeBreak(messageLengthBytes, "Node in-flight requests");
+            transportChannel.setReservedBytes(messageLengthBytes);
```

Comment:
I looked at this more closely. First of all it is possible to make both transport channel implementations immutable. However, in both transport cases the client does not get a response when the circuit breaker trips:

* LocalTransport: I think there is a problem in the catch block of `LocalTransport#messageReceived()`. If an exception is thrown, `transportServiceAdapter` does not have handler registered for the given request. The handler is just registered on the *sending* node but never on the *receiving* node (where the exception occurs). In the end, the exception is silently dropped and the client never gets the exception.
* NettyTransport: In this case, the exception ends up in `NettyTransport#exceptionCaught()` which is just closing the connection on exception and the client also never get hold of the exception.

So I propose that we implement handleRequest in both cases as follows (sketch):

```java
//imagine here local or netty transport
TransportChannel transportChannel = null; 
try {
  inFlightRequestsBreaker().addEstimateBytesAndMaybeBreak(messageLengthBytes, "<transport_request>");
  transportChannel = new TransportChannel(..., messageLengthBytes);
  // further request handling logic here
} catch (Throwable t) {
  // this can just happen if the circuit breaker tripped
  // create a transport channel which does not have to free any bytes
  if (transportChannel == null) {
  	transportChannel = new TransportChannel(..., 0);
  }
  transportChannel.sendResponse(e);
}
```

In that case we always send a response. Wdyt?

However, I think we have to check the case `LocalTransport#messageReceived()` (but I'd rather not do this in the scope of this ticket).